File: src-tauri/src/domain/models/word.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Word {
    pub id: String,
    pub word: String,
    pub translation: String,
    #[serde(default)]
    pub phonetic: Option<String>,
    #[serde(default = "default_source_lang")]
    pub source_lang: String,
    #[serde(default = "default_target_lang")]
    pub target_lang: String,
    #[serde(default = "default_status")]
    pub status: String,
    #[serde(default)]
    pub tags: String,
    #[serde(default)]
    pub notes: String,
    #[serde(default)]
    pub audio_url: Option<String>,
    #[serde(default = "default_ease_factor")]
    pub ease_factor: f64,
    #[serde(default)]
    pub interval: i32,
    #[serde(default)]
    pub repetitions: i32,
    #[serde(default)]
    pub next_review_at: Option<i64>,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct WordMeaning {
    pub part_of_speech: String,
    pub translations: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct EnglishDefinitionGroup {
    pub part_of_speech: String,
    pub definitions: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct WordFormItem {
    pub r#type: String,
    pub words: Vec<String>,
}

pub const WORD_FORM_TYPES: &[&str] = &[
    "lemma",
    "lemma_variant",
    "past_tense",
    "past_participle",
    "present_participle",
    "third_person_singular",
    "comparative",
    "superlative",
    "plural",
];

pub fn is_valid_word_form_type(value: &str) -> bool {
    WORD_FORM_TYPES.contains(&value)
}

fn default_source_lang() -> String {
    "EN".to_string()
}

fn default_target_lang() -> String {
    "ZH".to_string()
}

fn default_status() -> String {
    "new".to_string()
}

fn default_ease_factor() -> f64 {
    2.5
}

impl Word {
    pub fn new(
        id: String,
        word: String,
        translation: String,
        source_lang: String,
        target_lang: String,
    ) -> Self {
        let now = chrono::Utc::now().timestamp_millis();
        Word {
            id,
            word,
            translation,
            phonetic: None,
            source_lang,
            target_lang,
            status: "new".to_string(),
            tags: String::new(),
            notes: String::new(),
            audio_url: None,
            ease_factor: 2.5,
            interval: 0,
            repetitions: 0,
            next_review_at: None,
            created_at: now,
            updated_at: now,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn is_valid_word_form_type_accepts_known_values() {
        for kind in WORD_FORM_TYPES {
            assert!(is_valid_word_form_type(kind));
        }
    }

    #[test]
    fn is_valid_word_form_type_rejects_unknown_values() {
        assert!(!is_valid_word_form_type("abbrev"));
        assert!(!is_valid_word_form_type(""));
    }
}

File: src/lib/api/word.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  EnglishDefinitionGroup,
  WordDetail,
  WordFormItem,
} from "./wordDetails";
import { toWordDetail } from "./wordDetails";
import type { TranslationExample } from "./translate";

export interface WordMeaning {
  partOfSpeech: string;
  translations: string[];
}

export interface Word {
  id: string;
  word: string;
  translation: string;
  phonetic?: string;
  source_lang: string;
  target_lang: string;
  status: string;
  tags: string;
  notes: string;
  audio_url?: string;
  ease_factor: number;
  interval: number;
  repetitions: number;
  next_review_at: number;
  created_at: number;
  updated_at: number;
}

export interface WordUpdate {
  translation?: string;
  tags?: string;
  notes?: string;
  status?: string;
}

export interface AddWordInput {
  word: string;
  translation: string;
  sourceLang: string;
  targetLang: string;
  phonetic?: string | null;
  meanings: WordMeaning[];
  englishDefinitions: EnglishDefinitionGroup[];
  examples: TranslationExample[];
  wordForms: WordFormItem[];
  memoryTip: string;
  tags?: string;
}

export async function addWord(input: AddWordInput): Promise<WordDetail> {
  const detail = await invoke<Parameters<typeof toWordDetail>[0]>("add_word", {
    input,
  });
  return toWordDetail(detail);
}

export async function getWords(search?: string): Promise<Word[]> {
  return invoke("get_words", { search });
}

export async function deleteWord(wordId: string): Promise<void> {
  return invoke("delete_word", { wordId });
}

export async function updateWord(
  wordId: string,
  updates: WordUpdate,
): Promise<Word> {
  return invoke("update_word", { wordId, updates });
}
